### ISSUE TITLE
Glob and drive letter on Windows

### ISSUE THREAD

**Issue by melMass** [nushell/nushell#7125]
### Describe the bug

Glob doesn't seem to work with drive letters on Windows.

### How to reproduce

Try any of the following on Windows:

```nu
# Using backslash
glob C:\Users\User\AppData\Roaming\nushell\scripts\apps\*
# Using forward slash
glob C:/Users/User/AppData/Roaming/nushell/scripts/apps/*
# Constructed from nu pipelines
glob $"($config_dir | path join 'apps')(char path_sep)*"
# Constructed from nu pipelines sanitized
$"($config_dir | path join 'apps')(char path_sep)*" | str replace -a '\\' '/' | path expand | glob $in
```

```nu
# Let's try simpler alts

glob "C:/*"
glob "C:\*"
glob "C:*"
```

All of the these return

```ruby
Error:
  × error with glob pattern
  help: failed to parse glob expression
```

### Expected behavior

To work as expected with absolute paths on windows, including drive letter.

### Configuration

| key                | value                                           |
| ------------------ | ----------------------------------------------- |
| version            | 0.71.0                                          |
| branch             |                                                 |
| commit_hash        |                                                 |
| build_os           | windows-x86_64                                  |
| build_target       | x86_64-pc-windows-msvc                          |
| rust_version       | rustc 1.66.0-nightly (5c8bff74b 2022-10-21)     |
| rust_channel       | nightly-x86_64-pc-windows-msvc                  |
| cargo_version      | cargo 1.66.0-nightly (3ff044334 2022-10-17)     |
| pkg_version        | 0.71.0                                          |
| build_time         | 2022-11-13 19:16:05 +01:00                      |
| build_rust_channel | release                                         |
| features           | database, dataframe, default, trash, which, zip |
| installed_plugins  | iselect |

**Comment by fdncred** [comment]
Our `glob` command uses the `wax` crate so this may be an upstream problem but I'm not positive.

**Comment by melMass** [comment]
> `wax` crate so this may be an upstream problem but I'm not positive.

Ah yes most probably, I quickly looked at it and it seems that the `:` on windows might interfere with their `repetition` feature.
I'll open an issue there. Btw I think it should be specified somewhere in the doc, there are some nice tips on their readme!

**Comment by webbedspace** [comment]
[Verdict: WONTFIX](https://github.com/olson-sean-k/wax/issues/34)

**Comment by theAkito** [comment]
> [Verdict: WONTFIX (more or less)](https://github.com/olson-sean-k/wax/issues/34)

Why, though?

> ### Schemes and Prefixes
> 
> While globs can be rooted, they cannot include schemes nor Windows path
> prefixes. For example, the Windows UNC share path `\\server\share\src` cannot be
> represented directly as a glob.
> 
> This can be limiting, but the design of Wax explicitly forbids this: Windows
> prefixes and other volume components are not portable. Instead, when this is
> needed, an additional native path or working directory must be used, such as
> [the `--tree` option provided by Nym][nym]. In most contexts, globs are applied
> relative to some such working directory.

##### [Source](https://github.com/olson-sean-k/wax#schemes-and-prefixes)

This and the comment provided by the maintainer display possibilities to fix this issue, even when they aren't perfect or "portable" in the sense described.

**Comment by webbedspace** [comment]
Sorry, I meant the upstream bug with Wax is WONTFIX, not this one. This one can presumably be fixed with a little bit of unattractive D.I.Y.

**Comment by theAkito** [comment]
> Sorry, I meant the upstream bug with Wax is WONTFIX, not this one. This one can presumably be fixed with a little bit of unattractive D.I.Y.

Ah yes, nice.

**Comment by fdncred** [comment]
There's also been talk about removing `wax` and `glob` entirely. I'd prefer to keep it and expand on the `glob` functionality since I think it's a good add for a composable pipeline. The reason why I added `wax` was because `nu-glob` had to have some globbing functionality from `nu-glob` because it was breaking `ls` is some weird way. I can't remember why but JT removed it/tweaked it a long time ago. One could probably look into the PRs and find out why.

**Comment by olson-sean-k** [comment]
Maintainer of `wax` here. I'm happy to participate, so I thought I'd comment with a few thoughts.

> because [`nu-glob`] was breaking `ls` is some weird way.

Perhaps #5691?

> I think it's a good add for a composable pipeline.

I agree. While it is more verbose, I find the idea of a dedicated `glob` command more clear and consistent and  I think `wax` could be a good fit for this (despite needing an additional mechanism for platform-specific path features). This may be an unpopular opinion, but I think something like `glob -a *.txt | rm`<sup>1</sup> is preferable to `rm -a *.txt`. Going even further, commands (or the shell itself) need not support globbing at all. This would make globbing and path operations more obvious, keep globbing features contained to the `glob` command, and make documentation easier to find and read (it is more obvious where to look for globbing documentation and command documentation is less cluttered).

Regarding globs as an integrated shell feature, native paths with pattern extensions are the norm and users may be confused or otherwise react negatively to a departure from that, so it's a bit less clear to me if `wax` is suitable as is. It would probably require the same mechanism for platform-specific path features in every command that supports globbing.

<sup>1</sup>_Assuming `rm` can accept paths from standard input. An expansion or `xargs` equivalent could also do._

**Comment by theAkito** [comment]
> This may be an unpopular opinion, but I think something like `glob -a *.txt | rm`1 is preferable to `rm -a *.txt`.

I think, this seems very reasonable in the context of Nu. 👍🏻



> Regarding globs as an integrated shell feature, native paths with pattern extensions are the norm and users may be confused or otherwise react negatively to a departure from that, so it's a bit less clear to me if `wax` is suitable as is.

In my view, it's important to question the norm. Especially, when talking about OS stuff. Too much stuff did not change for decades, just because "we have always done it like this", which is an atrocious reason for keeping something, as it is. My 2 cents on that topic, at least.



> It would probably require the same mechanism for platform-specific path features in every command that supports globbing.

I don't imagine this too require too much effort or ugliness. Correct me, if I'm wrong.

**Comment by fdncred** [comment]
@olson-sean-k Thanks for chiming in.

> Perhaps https://github.com/nushell/nushell/pull/5691?

No, that's not the one I was referring to. I think it's this one https://github.com/nushell/nushell/pull/4818

**Comment by olson-sean-k** [comment]
> I don't imagine this too require too much effort or ugliness. Correct me, if I'm wrong.

I don't believe it would. An option or positional argument for a working directory as a native path should be enough, perhaps something like `glob --tree=A:\ **/*.txt`. In the implementation, the default value for this option would be the current working directory and it would be used as a parameter in calls to `wax::Glob::walk`.

On Unix platforms, this will almost never come up since paths are comparatively simple. On Windows, it may be necessary to change directory prior to a command or to use this option to target a specific volume (drive letter, UNC share, verbatim path, etc.).

**Comment by melMass** [comment]
Interesting. 
All these discussions led me to try and just replace every `glob` instances with `(ls <globpatter>).name` and **ALL** the samples I posted in the OP are now working. I guess it's a design decision now regarding what was bring up

**Comment by olson-sean-k** [comment]
> ...replace every `glob` instances with `(ls <globpatter>).name` and **ALL** the samples I posted in the OP are now working.

Yes, that sounds right. I believe that commands like `ls` have direct support for globbing using a different in-tree library. That library always parses globs as native paths (via `std::path`), ignores any prefix and root components, and then extracts patterns from the remaining components. This approach has different tradeoffs than that of Wax and allows for native paths with pattern extensions (there's no need to isolate platform-specific path features for the most part).

**Comment by olson-sean-k** [comment]
Hi again, folks! I've become aware of several Nushell issues that seem to relate to this one and there is [a similar issue](https://github.com/olson-sean-k/wax/issues/34) open against `wax`. I'd like to share some thoughts and crazy ideas and thought this may be a good place to do so: please let me know if there's a better place though!

I'd really like to avoid introducing platform-specific features to glob expressions in `wax`. One tenant of that library is that glob expressions are portable and distinct from paths (and moreover are not "extended native paths"). `wax` glob expressions are the same everywhere. This doesn't always work so well for the problems faced by a shell though.

This is a big departure from the way this is usually done, but I think the way that I'd prefer to address this problem is to provide a mechanism atop `wax` that allows native paths and patterns to be used together and yet remain distinct. I think @bobhy suggested something similar [here](https://github.com/olson-sean-k/wax/issues/34#issuecomment-1741865497), for example.

It needs more thought, but I've put together a quick proof of concept called [treeish](https://github.com/olson-sean-k/treeish). That library provides a `Treeish` type, which joins native path prefixes and globs into a single expression. That repository includes an example CLI tool called `findish` that accepts a treeish and prints the paths of files with a matching path that it finds. A treeish looks something like the following and uses an optional `::` separator in its explicit form:

- `**/*.txt`
- `\\?\UNC\server\share::**/*.txt`
- `C:\Users::**/AppData/**/*.savedgame`
- `A:\novel.txt`
- `\\?\GLOBALROOT\Device::HarddiskVolume<[0-9]:1,>`

A `Treeish` can be one of three things:

1. A native path. The parser prefers a native path interpretations in general.
1. A glob pattern.
1. A glob pattern in a particular tree (i.e., a native path prefix and a glob).

Rather than accepting or constructing paths (i.e., `std::path::Path{Buf}`), CLI programs can instead accept or construct a `Treeish` for any argument that should support patterns. `Treeish` may even help programs provide informative errors about arguments that should **not** accept patterns by detecting and rejecting them (at time of writing, via `Treeish::new(...)?.path()`, which return `None` if any patterns are parsed). Because `treeish` is just a library, it should be possible for applications that aren't associated with Nushell to provide a similar interface too.

I just wanted to mention this idea in case anyone has thoughts or opinions about this kind of approach. It's a bit unusual, but I really prefer distinguishing between paths (prefixes) and patterns. For better or worse, users on Unix may miss this altogether while users on Windows may see `::` as a strange and unfair annoyance. I'm not sure. :smile: I may be off my rocker here.

To play with `findish`, clone the `treeish` repository and try `cargo run -p findish -- '<treeish>'` (note that the treeish expression is quoted to avoid escaping).

**Comment by bobhy** [comment]
Speaking just for myself and in no way for the rest of the Nu community...
I doubt any kind of explicit partitioning syntax would get much traction in Nu, whether it's within the pattern as above or as a separate argument (like `glob --root <abspath> <pattern>`) -- it would be a departure from globbing in other shells and would only be helpful to a very small minority of users, to wit:
* I want to enumerate files, e.g to iterate through them, doing something with each one (perhaps 100% of users want to do this at some point)
* I don't think of or choose to use  `ls` to list the files.  It can  filter by type as can `glob`, but can also sort by size or date. I think this makes it the more natural first choice, so I guess only 10% of users fall through)
* I'm running on Windows. (I guess 40% of Nu users, but would be interested in hard data)
* The files I'm interested in don't reside in or below CWD. (say 75% of users do glob outside of CWD)
* The files can't conveniently or naturally be referenced with a CWD-relative pattern.  We recently merged #10504, to enable `glob ../<pattern>`.  I'm guessing that covers most user needs, but say 25% of users fall through)

This multiplies down to an (arguably generous) estimate of 0.75% of nu users who would need to discover and use whatever alternative syntax `glob` provides.  Is the game worth the candle?

**Comment by fdncred** [comment]
It seems to me that the hardest part of this problem is figuring out which part is a glob and which part is a path. Since MacOS & Linux can have glob patterns in their path and file names, I'm not sure there's a good way to determine this automatically. I'm definitely interested in automatic ways to do this, but in the meantime, I'm interested in @olson-sean-k's prototype solution. I plan on taking a look at it because it takes the guesswork out of it and specifically requires you to opt-in to "here's the path" and "here's the glob". Not my first choice at all, but in the absence of an automatic way to do it right every time, I may use it.

Thanks @olson-sean-k. I look forward to playing with your software more and much thanks for `wax`. I was super pumped to see you supporting it again. It's a great globbing crate.... except for Windows users. 😉 LOL.
